src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs

```diff
@@ -0,0 +1,210 @@
+        public object Property { get; set; }
+
+        /// <summary>
+        /// Gets or sets the delimiter to join the output with.
+        /// </summary>
+        [Parameter(Position = 1)]
+        [ArgumentCompleter(typeof(JoinItemCompleter))]
+        [AllowEmptyString]
+        public string Delimiter { get; set; }
+
+        /// <summary>
+        /// Gets or sets text to include before the joined input text.
+        /// </summary>
+        [Parameter]
+        public string Prefix { get; set; }
+
+        /// <summary>
+        /// Gets or sets text to include after the joined input text.
+        /// </summary>
+        [Parameter]
+        public string Suffix { get; set; }
+
+        /// <summary>
+        /// Gets or sets if the output items should we wrapped in single quotes.
+        /// </summary>
+        [Parameter(ParameterSetName = "Quote")]
+        public SwitchParameter Quote { get; set; }
+
+        /// <summary>
+        /// Gets or sets if the output items should we wrapped in double quotes.
+        /// </summary>
+        [Parameter(ParameterSetName = "DoubleQuote")]
+        public SwitchParameter DoubleQuote { get; set; }
+
+        /// <summary>
+        /// Gets or sets the input object to join into text.
+        /// </summary>
+        [Parameter(ValueFromPipeline = true, Mandatory = true)]
+        public PSObject InputObject { get; set; }
+
+        /// <inheritdoc />
+        protected override void ProcessRecord()
+        {
+            if (InputObject != null && InputObject != AutomationNull.Value)
+            {
+                _inputObjects.Add(InputObject);
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void EndProcessing()
+        {
+            if (_inputObjects.Count == 0)
+            {
+                return;
+            }
+
+            var quoteChar = Quote ? '\'' : DoubleQuote ? '"' : char.MinValue;
+
+            const int DefaultOutputStringCapacity = 256;
+            var builder = new StringBuilder(DefaultOutputStringCapacity);
+            builder.Append(Prefix);
+
+            if (Delimiter == null)
+            {
+                Delimiter = LanguagePrimitives.ConvertTo<string>(GetVariableValue("OFS"));
+            }
+
+            void AppendValue(string val)
+            {
+                if (quoteChar != char.MinValue)
+                {
+                    builder.Append(quoteChar).Append(val).Append(quoteChar);
+                }
+                else
+                {
+                    builder.Append(val);
+                }
+            }
+
+            if (Property == null)
+            {
+                AppendValue(LanguagePrimitives.ConvertTo<string>(_inputObjects[0]));
+
+                for (var index = 1; index < _inputObjects.Count; index++)
+                {
+                    builder.Append(Delimiter);
+                    AppendValue(LanguagePrimitives.ConvertTo<string>(_inputObjects[index]));
```

Comment:
we do not need to collect input objects - we need only result strings. So we should move the ConvertTo to the ProcessRecord() method.